compiler/rustc_const_eval/src/interpret/intrinsics.rs

```diff
@@ -573,6 +601,106 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
         })
     }
 
+    pub fn carrying_arith(
+        &self,
+        mir_op: BinOp,
+        l: &ImmTy<'tcx, M::Provenance>,
+        r: &ImmTy<'tcx, M::Provenance>,
+        c: &ImmTy<'tcx, M::Provenance>,
+    ) -> InterpResult<'tcx, (Scalar<M::Provenance>, Scalar<M::Provenance>)> {
+        assert_eq!(l.layout.ty, r.layout.ty);
+        assert_matches!(l.layout.ty.kind(), ty::Int(..) | ty::Uint(..));
+        assert_matches!(c.layout.ty.kind(), ty::Bool);
+        assert_matches!(mir_op, BinOp::Add | BinOp::Sub);
+
+        let mir_op = mir_op.wrapping_to_overflowing().unwrap();
+
+        let (val, overflowed1) = self.binary_op(mir_op, l, r)?.to_scalar_pair();
+
+        let val = ImmTy::from_scalar(val, l.layout);
+        let c = ImmTy::from_scalar(c.to_scalar(), l.layout);
+
+        let (val, overflowed2) = self.binary_op(mir_op, &val, &c)?.to_scalar_pair();
+
+        let overflowed1 = overflowed1.to_bool()?;
+        let overflowed2 = overflowed2.to_bool()?;
+
+        let overflowed = Scalar::from_bool(if l.layout.abi.is_signed() {
+            overflowed1 != overflowed2
+        } else {
+            overflowed1 | overflowed2
+        });
```

Comment:
Please add a comment explaining the logic here

Reply:
I definitely should just update the intrinsic docs to clarify the behaviour here, although the standard library has a bit of a weird relationship with how it documents intrinsics.

Most of them rely on having stabilised versions that they can just point to, since those will have the proper docs. The documentation here is split between `iN::carrying_add`, `uN::carrying_add`, `iN::borrowing_sub`, and `uN::borrowing_sub`, but the bottom line is that signed methods are merely checking for overflow, whereas unsigned methods want to actually return a new carry bit that can be chained along. That's what we're testing for in the methods and I'm just duplicating that here.

Not sure what the best solution for documentation would be here; open to ideas. I could just link those docs here, for now.

Reply:
I just don't understand why the signed thing overflowed if exactly one of the sub-operations overflowed. Like I could probably think about it for a minute and figure it out, but there should really be comments explaining that.

Your answer confused me even more, in which sense is the result different for signed vs unsigned? That should also be documented...